include/exec/__detail/__system_context_default_impl_entry.hpp

```diff
@@ -23,23 +23,111 @@
 
 #include "__system_context_default_impl.hpp"
 
-STDEXEC_PRAGMA_PUSH()
-STDEXEC_PRAGMA_IGNORE_GNU("-Wattributes") // warning: inline function '[...]' declared weak
-
-/// Gets the default system context implementation.
-extern "C" STDEXEC_SYSTEM_CONTEXT_INLINE STDEXEC_ATTRIBUTE((weak))
-__exec_system_context_interface*
-  __get_exec_system_context_impl() {
-  return exec::__system_context_default_impl::__instance_holder::__singleton()
-    .__get_current_instance();
-}
-
-/// Sets the default system context implementation.
-extern "C" STDEXEC_SYSTEM_CONTEXT_INLINE STDEXEC_ATTRIBUTE((weak))
-  void
-  __set_exec_system_context_impl(__exec_system_context_interface* __instance) {
-  return exec::__system_context_default_impl::__instance_holder::__singleton()
-    .__set_current_instance(__instance);
-}
-
-STDEXEC_PRAGMA_POP()
+#include <atomic>
+#include <thread>
+
+namespace exec {
+  namespace __detail {
+    STDEXEC_SYSTEM_CONTEXT_INLINE constinit __spin_lock __system_context_mutex{};
+
+    STDEXEC_SYSTEM_CONTEXT_INLINE constinit std::atomic<system_context_interface*>
+      __system_context_instance{nullptr};
+
+    STDEXEC_SYSTEM_CONTEXT_INLINE constinit std::atomic<new_system_context_handler>
+      __new_system_context_handler{[]() -> system_context_interface* {
+        return new __detail::__system_context_impl{};
+      }};
+
+    // The job of this object is to hold a reference to whatever object
+    // is the current system context from the time the first system
+    // context object is created until after main exits.
+    struct __system_context_keep_alive {
+      __system_context_keep_alive() {
+        (void) get_system_context_instance();
+      }
+
+      ~__system_context_keep_alive() {
+        release_system_context_instance(__system_context_instance.load());
+      }
+    };
+
+    struct __system_context_base {
+      __system_context_base() {
+        // This static will create the system context, and decrement its
+        // ref count after main exits.
+        static __system_context_keep_alive __s_keep_alive;
+      }
+    };
```

Comment:
The `__system_context_base` class doesn't seem to be used anywhere. What is it's purpose?

Also, the `__system_context_keep_alive` class is only used by `__system_context_base` and so seems to be dead-code.

Reply:
That was an oversight. `system_context` should inherit from `__system_context_base`. I have fixed that.